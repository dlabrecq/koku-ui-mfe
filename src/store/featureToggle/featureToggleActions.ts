import { createAction } from 'typesafe-actions';

export interface FeatureToggleActionMeta {
  isDebugFlagEnabled?: boolean;
  isUtilizationFlagEnabled?: boolean;
}

export const setFeatureToggle = createAction('feature/init_feature_toggle')<FeatureToggleActionMeta>();
export const resetState = createAction('feature/reset_state')();
