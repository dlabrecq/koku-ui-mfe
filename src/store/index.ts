export { configureStore, mfeStore } from './store';
export type { RootState } from './rootReducer';
