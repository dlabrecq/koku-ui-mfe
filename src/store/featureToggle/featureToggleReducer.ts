import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import type { resetState } from './featureToggleActions';
import { setFeatureToggle } from './featureToggleActions';

export type FeatureToggleAction = ActionType<typeof setFeatureToggle | typeof resetState>;

export type FeatureToggleState = Readonly<{
  hasFeatureToggle: boolean;
  isDebugFlagEnabled: boolean;
  isUtilizationFlagEnabled: boolean;
}>;

export const defaultState: FeatureToggleState = {
  hasFeatureToggle: false,
  isDebugFlagEnabled: false,
  isUtilizationFlagEnabled: false,
};

export const stateKey = 'featureToggle';

export function featureToggleReducer(state = defaultState, action: FeatureToggleAction): FeatureToggleState {
  switch (action.type) {
    case getType(setFeatureToggle):
      return {
        ...state,
        hasFeatureToggle: true,
        isDebugFlagEnabled: action.payload.isDebugFlagEnabled,
        isUtilizationFlagEnabled: action.payload.isUtilizationFlagEnabled,
      };

    default:
      return state;
  }
}
