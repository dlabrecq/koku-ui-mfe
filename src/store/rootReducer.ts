import { notificationsReducer } from '@redhat-cloud-services/frontend-components-notifications/redux';
import { combineReducers } from 'redux';
import { accountSettingsReducer, accountSettingsStateKey } from 'store/accountSettings';
import { forecastReducer, forecastStateKey } from 'store/forecasts';
import { reportReducer, reportStateKey } from 'store/reports';
import { resourceReducer, resourceStateKey } from 'store/resources';
import { rosReducer, rosStateKey } from 'store/ros';
import type { StateType } from 'typesafe-actions';

import { featureFlagsReducer, featureFlagsStateKey } from './featureFlags';
import { providersReducer, providersStateKey } from './providers';
import { uiReducer, uiStateKey } from './ui';
import { userAccessReducer, userAccessStateKey } from './userAccess';

export type RootState = StateType<typeof rootReducer>;

export const rootReducer = combineReducers({
  [accountSettingsStateKey]: accountSettingsReducer,
  [featureFlagsStateKey]: featureFlagsReducer,
  [forecastStateKey]: forecastReducer,
  [providersStateKey]: providersReducer,
  [reportStateKey]: reportReducer,
  [resourceStateKey]: resourceReducer,
  [rosStateKey]: rosReducer,
  [uiStateKey]: uiReducer,
  [userAccessStateKey]: userAccessReducer,
  notifications: notificationsReducer,
});
