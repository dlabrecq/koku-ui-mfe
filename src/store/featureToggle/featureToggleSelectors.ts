import type { RootState } from 'store/rootReducer';

import { stateKey } from './featureToggleReducer';

export const selectFeatureToggleState = (state: RootState) => state[stateKey];

export const selectHasFeatureToggle = (state: RootState) => selectFeatureToggleState(state).hasFeatureToggle;

export const selectIsDebugFlagEnabled = (state: RootState) => selectFeatureToggleState(state).isDebugFlagEnabled;
export const selectIsUtilizationFlagEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isUtilizationFlagEnabled;
