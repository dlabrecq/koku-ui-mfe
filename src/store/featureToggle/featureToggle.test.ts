import { createMockStoreCreator } from 'store/mockStore';

import { featureToggleSelectors } from '.';
import * as actions from './featureToggleActions';
import { featureToggleReducer, stateKey } from './featureToggleReducer';
import * as selectors from './featureToggleSelectors';

const createUIStore = createMockStoreCreator({
  [stateKey]: featureToggleReducer,
});

test('default state', async () => {
  const store = createUIStore();
  expect(selectors.selectFeatureToggleState(store.getState())).toMatchSnapshot();
});

test('Utilization feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isBoxPlotToggleEnabled: true }));
  expect(featureToggleSelectors.selectIsBoxPlotToggleEnabled(store.getState())).toBe(true);
});

test('Project link feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isProjectLinkToggleEnabled: true }));
  expect(featureToggleSelectors.selectIsProjectLinkToggleEnabled(store.getState())).toBe(true);
});
