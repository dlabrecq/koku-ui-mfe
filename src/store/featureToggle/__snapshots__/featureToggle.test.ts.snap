// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
{
  "hasFeatureToggle": false,
  "isBoxPlotToggleEnabled": false,
  "isDebugToggleEnabled": false,
  "isProjectLinkToggleEnabled": false,
}
`;
